test/hotspot/jtreg/runtime/cds/appcds/jvmti/CFLH/MultiReleaseJars.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @summary Test multi-release jar with CFLH
+ * @requires vm.cds
+ * @requires vm.jvmti
+ * @library /test/lib /test/hotspot/jtreg/runtime/cds/appcds
+ * @run main/othervm/native MultiReleaseJars
+ */
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.PrintStream;
+import java.io.IOException;
+import jdk.test.lib.cds.CDSTestUtils;
+import jdk.test.lib.process.OutputAnalyzer;
+
+public class MultiReleaseJars {
+
+    static final int BASE_VERSION = 9;
+    static final String BASE_VERSION_STRING = Integer.toString(BASE_VERSION);
+    static final int MAJOR_VERSION = Runtime.version().major();
+    static final String MAJOR_VERSION_STRING = String.valueOf(MAJOR_VERSION);
+
+    static String[] getMain() {
```

Comment:
Can't you just define these as source files in the test directory rather than creating them dynamically?

Reply:
I think the concern may be that we want to avoid license headers.

However, we should still use text blocks instead of these string arrays for such inlined source files.

Reply:
I've changed the code to use text blocks.

Reply:
We have two versions of the "Foo" class, but jtreg doesn't support building of classes with the same name. If you look at other multi-release JAR test cases, they all have to build the classes by hand. Embedded the Java sources in the test case is actually simpler than these existing tests:

- https://github.com/openjdk/jdk/blob/6f622da7fbae67d8c1cd9e795127adac58a246a9/test/jdk/tools/jar/multiRelease/MRTestBase.java
- https://github.com/openjdk/jdk/blob/6f622da7fbae67d8c1cd9e795127adac58a246a9/test/jdk/tools/jar/multiRelease/Basic.java